### ISSUE TITLE
Some problem about the defence detction

### ISSUE THREAD

**Issue by RandolphCarter0** [Megum1/LOTUS#1]
Hi, LOTUS is an interesting work, thank you for sharing! 
I tried to reproduce the LOTUS performance of the defense method ```neural cleanse``` (settings referenced to their [code)](https://github.com/bolunwang/backdoor) according to the configuration in the LOTUS paper. But I found that the norm values on resnet18 are **NAN** for all the class labels. Did I do something wrong during the operation (Maybe I need to change the configuration of resnet18_lotus? or settings of neural cleanse? )? May I ask you to release more details related to the backdoor defence experiment?
![20240726181456](https://github.com/user-attachments/assets/7a85ddf8-cdeb-4efb-bac7-06e37e203d25)

**Comment by Megum1** [comment]
Thank you for your interest in our work!
The issue with the norm values being NaN may be due to the configuration/environment inconsistencies.
To help you better, I’ve attached our Neural Cleanse code in the `misc/nc.py` file. Hope that could be helpful.

**Comment by RandolphCarter0** [comment]
Thank you for your sharing!
I tried to run the code in the file `misc/nc.py`, with the set of resnet18 in the paper, and it's shown that the method lotus had a good performance in the detection algorithms like neural cleanse!
But I found an interesting phenomenon in the experiments. I used the nc method to make detection for a clean model (resnet18 in cifar10) trained in step 1 in file `main.py` and got the result of the anomaly index which is **2.72** for repeated experiments. Is it related to the early stop mechanism and MAD outlier detection algorithm in nc or other configuration inconsistencies?  I'm looking forward to hearing your opinion.

**Comment by Megum1** [comment]
Hi. Thanks for you interest and this is a good observation!!

Originally, NC is designed to detect the universal backdoor (victim can be any sample) and it works well. So if you run `python nc.py --model_filepath checkpoint/clean.pt --victim -1`. It should output an anomaly score as ~0.8.

In our experiment, to enhance NC's ability to detect label-specific backdoor (victim samples should be from the victim class), we only feed validation samples from the victim class and calculate the anomaly score. This could introduce false positive cases, where a clean model is detected as backdoor.

We suspect there are several potential reasons for the false positive case.
(1) The clean model itself may have natural vulnerability after training. For example, it is very easy to flip samples from Class dog to Class cat using a small adversarial perturbation. This is possible as the typical training framework does not consider robust adversarial training.
(2) The samples we select for NC are not confident. For example, some dog samples have 0.5 confidence on Class dog and ~0.3 confidence on Class cat and hence it's easy to flip the prediction.
(3) As you have mentioned, NC's early stop mechanism may stop the optimization before it is optimal or fully converged. For example, the `patience=5` value is a bit small and NC will stop when the mask size is not reduced for 5 steps, but it may be reduced at the 6th step.

I think a possible and easiest solution to mitigate the false positive cases is to try different random seeds `--seed` and probably set a high `patience` such that NC could have more steps to refine its output.

Hope that could help and thanks again for your comments.

**Comment by RandolphCarter0** [comment]
Thank you for your profound insights！ Now the Lotus attack works well on the detection method! Inspired by the impressive and interesting attack Lotus, we'll further explore the backdoor attack and defense.

Meanwhile, I noted that Lotus performed excellently on the backdoor mitigation method. May I ask you to release more detail of the experiment set for the backdoor mitigation method (in the paper Table 3) such as the source code for FP, ANP, and NAD? I would be extremely grateful for your reply.

**Comment by Megum1** [comment]
Hi, thanks again for your continued interest! We are planning to release our mitigation method in the near future.
In the meantime, you can check out another repository of ours that contains implementations of recent mitigation techniques: [OrthogLinearBackdoor](https://github.com/KaiyuanZh/OrthogLinearBackdoor/blob/main/backdoor_mitigation.py)

**Comment by RandolphCarter0** [comment]
Thank you for your good advice! The view of the paper you mentioned is also interesting and pioneering. Your comments solve my problem. I'm looking forward to more backdoor attack work from you, thank you again! I'll close the question later.

**Comment by RandolphCarter0** [comment]
I'm sorry to open this issue again. I explored the backdoor mitigation method in Lotus according to the reference OB [OrthogLinearBackdoor](https://github.com/KaiyuanZh/OrthogLinearBackdoor/blob/main/backdoor_mitigation.py) (That's also an impressive work! Thank for your sharing again!).

According to the implementation of the project and the settings mentioned in the Lotus paper, I test the Lotus method in the four mitigation methods shown in Table 3. I find problems as follows: (1) With the set in OB, the fine-tune (ft) method works well and is close to the result in the paper, while the fineprune (fp) method gets the asr 27~30% in lotus and is lower than the result in the paper. (2) the result of ANP and NAD methods differs from that in Table 3 with repeated experiments. 

I attribute the difference to the reason that the settings of these methods are ambiguous (e.g. the rate of prune in the fp methods in lotus is unknown). May I request you to tell me more details about the settings of mitigation methods in section 5.4 of the paper (especially the value of hyperparameters for these methods)?

Besides, I have another question: can we regard the Lotus method as a special target Badnet method when the number of the partition is one?

The Lotus method is impressive, thank you for sharing again! I would be grateful if you could reply to me. Thank you again!

**Comment by Megum1** [comment]
Hi. Thanks for revisiting this issue and for your valuable comments!

Regarding your observations, we want to clarify a few points. In our experiments, we used specific hyperparameter settings for the baseline methods to balance backdoor removal and maintain clean accuracy. For instance, in the fine-prune (fp) method, parameters like the pruning rate were carefully chosen. If higher pruning rates or learning rates are applied, it’s expected that the ASR could be reduced further, but this often comes at the cost of clean accuracy. We apologize that we don’t have a record of the exact hyperparameter values used, but this trade-off is critical and may explain the differences in your results compared to the paper.

Additionally, to enhance the resilience of Lotus, we typically employ stronger adversarial training in trigger focusing. This strengthens the binding between the trigger and benign features, making the method more robust. Variations in results across experiments are understandable, as specific hyperparameter choices can affect outcomes. However, we believe that Lotus consistently outperforms existing methods, even if the degree of improvement varies based on the setup.

On your question about viewing Lotus as a targeted BadNet method when the partition count is one, this perspective is conceptually valid. A single partition simplifies the framework, resembling a targeted approach. However, the true strength of Lotus lies in its ability to handle multiple partitions effectively, distinguishing it from traditional methods.

We hope this addresses your questions. Thanks again for your interest and for sharing your observations.

**Comment by JoyRider0716** [comment]
@RandolphCarter0 

Hello, I am a novice and just started learning AI. I explored the backdoor mitigation method in Lotus with reference to OB [OrthogLinearBackdoor](https://github.com/KaiyuanZh/OrthogLinearBackdoor/blob/main/backdoor_mitigation.py). The results of the reproduction are quite different from the original results. Did you finally reproduce the results similar to the original results? If so, can you share more details?
